File: common/src/main/java/org/sandboxpowered/silica/resources/ResourceLoader.java
package org.sandboxpowered.silica.resources;

import org.sandboxpowered.api.util.Identity;

import java.io.IOException;
import java.io.InputStream;
import java.util.Set;
import java.util.function.Predicate;

public interface ResourceLoader {
    private static String getFilename(ResourceType type, Identity identity) {
        return String.format("%s/%s/%s", type.getFolder(), identity.getNamespace(), identity.getPath());
    }

    boolean containsFile(ResourceType type, String path);

    InputStream openFile(ResourceType type, String path) throws IOException;

    Set<String> findResources(ResourceType type, String namespace, String path, int depth, Predicate<String> filter);

    Set<String> getNamespaces(ResourceType type);
}

File: common/src/main/kotlin/org/sandboxpowered/silica/resources/ResourceLoader.kt
package org.sandboxpowered.silica.resources

import org.apache.commons.lang3.SystemUtils
import org.sandboxpowered.utilities.Identifier
import java.io.File
import java.io.InputStream
import java.util.function.Predicate

interface ResourceLoader {
    val name: String

    fun contains(type: ResourceType, file: Identifier): Boolean

    fun open(file: String): InputStream

    fun open(type: ResourceType, file: Identifier): InputStream

    fun findResources(
        type: ResourceType,
        namespace: String,
        path: String,
        depth: Int,
        filter: Predicate<String>
    ): Set<Identifier>

    fun getNamespaces(type: ResourceType): Set<String>

    companion object {
        fun getPath(type: ResourceType, identifier: Identifier): String {
            return "${type.folder}/${identifier.namespace}/${identifier.path}"
        }

        fun isValidPath(file: File, filename: String): Boolean {
            var string = file.canonicalPath
            if (SystemUtils.IS_OS_WINDOWS) {
                string = string.replace('\\', '/')
            }
            return string.endsWith(filename)
        }
    }
}